calyx-py/test/red_tree.py

```diff
@@ -0,0 +1,191 @@
+    It then runs the tree, and when the tree is done, stores the answer in {ans_reg}.
+    """
+    with comp.group(group) as tree_use:
+        a.addr0 = cb.const(32, a_i)
+        b.addr0 = cb.const(32, b_i)
+        c.addr0 = cb.const(32, c_i)
+        d.addr0 = cb.const(32, d_i)
+        tree.leaf1 = a.read_data
+        tree.leaf2 = b.read_data
+        tree.leaf3 = c.read_data
+        tree.leaf4 = d.read_data
+        tree.go = cb.const(1, 1)
+        # AM, point of failure: I need to add the following:
+        #   sum_col0.write_en = tree.done ? 1'b1;
+        #   sum_col0.in = tree.done ? tree.sum;
+        # But I don't know how to do that.
+        ans_reg.write_en = 1
+        ans_reg.in_ = tree.sum
+        tree_use.done = ans_reg.done
+
+
+def use_tree_ports_provided(comp, group, p1, p2, p3, p4, tree, ans_mem):
+    """Orchestrates the use of the component `tree`.
+    Adds wiring for {group}, which puts into the tree's four leaves
+    the values p1, p2, p3, and p4.
+    It then runs the tree, and stores the answer in the std_mem {ans_mem}.
+    """
+    # i.e., much like the above, but instead of calculating the
+    # ports, it takes them as arguments.
+
+    with comp.group(group) as tree_use:
+        tree.leaf1 = p1
+        tree.leaf2 = p2
+        tree.leaf3 = p3
+        tree.leaf4 = p4
+        tree.go = cb.const(1, 1)
+        # AM, point of failure:
+        # Just the like the point of failure above,
+        # I need the next three assignments to all be guarded by tree.done.
+        ans_mem.addr0 = cb.const(1, 0)
+        ans_mem.write_data = tree.sum
+        ans_mem.write_en = 1
+        tree_use.done = ans_mem.done
+
+
+def add_main(prog):
+    """Inserts the component `main` into the program.
+    This will be used in concert with multiple copies of the component `tree`.
+    It requires:
+    - Memories `A`, `B`, `C`, `D`, of length 4 each, to be driven from the data file.
+    - A memory `ans` to store the result, also driven from the data file.
+
+    It puts the sum of elements of the four memory cells into `ans`.
+    """
+    main = prog.component("main")
+    A = main.mem_d1("A", 32, 4, 32, is_external=True)
+    B = main.mem_d1("B", 32, 4, 32, is_external=True)
+    C = main.mem_d1("C", 32, 4, 32, is_external=True)
+    D = main.mem_d1("D", 32, 4, 32, is_external=True)
+    ans = main.mem_d1("ans", 32, 1, 1, is_external=True)
+    sum_col0 = main.reg("sum_col0", 32)
+    sum_col1 = main.reg("sum_col1", 32)
+    sum_col2 = main.reg("sum_col2", 32)
+    sum_col3 = main.reg("sum_col3", 32)
+
+    # AM, point of failure:
+    # I'd like to add the following to the `cells` section:
+    # tree0 = tree();
+    # I think the following is what you want me to do:
+    tree = main.cell("tree", prog.component("tree"))
+    # But _it adds a new, blank component called tree_ to the program.
+    # I'd like for it to locate the existing component `tree`.
+    # Thoughts?
```

Comment:
You will want to return the component definition from the `add_tree` function. Then, use that thing as the second argument to `main.cell`. Here's an example from my AOC work:
https://github.com/cucapra/aoc2022-calyx/blob/b90efb5e2181b08fd4c6a4a2998e5e23b21e0689/1/accelgen.py#L101-L103

Reply:
Woo nice! @EclecticGriffin this gets around the `CompInst("tree", [])` thing too! In fact this means that I don't need to import the Calyx AST at all, which is super!